### PR TITLE
Add support for terabytes in -X options

### PR DESCRIPTION
**Pull request by avermeer**
This fix adds support for 'T' and 't' suffix (indicating terabytes) to -X options that take a `<size>` parameter.

Fixes: #7427

### CODE REVIEW

**Comment by avermeer** [comment]
Made an additional PR to make sure new size suffix only applies to 64-bit versions of OpenJ9, with related unit tests.

**Comment by pshipton** [comment]
jenkins test sanity win,win32 jdk8

**Comment by pshipton** [comment]
@avermeer you can squash your commits and force push the result. We don't need to have 3 separate commits for this change.

**Comment by avermeer** [comment]
@pshipton here's my first attempt to squash some commits into one. I used this page as a clue, but I'm unsure of the result, because when I run `git log --oneline`, I see even more commits. At least the local files in my git directories still have the intended modifications.

As I must confess that I am more used to badminton than squash, I'd love to hear some feedback on whether or not my first squashing worked, and if not what should I have done to make it right...

**Comment by pshipton** [comment]
Ya, the squashing didn't work, there are now 5 commits but there should only be one. I can't help myself with the git command line, because I use Eclipse to interact with git, where there is a Modify -> Squash menu entry. Googling came up with https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/ as the first hit, if that helps.

**Comment by dmitripivkine** [comment]
I wondering `scan_udata_memory_size_helper()` in `mmparse.cpp` should be updated as well. This helper is used in parsing of GC related options

**Comment by avermeer** [comment]
Hi @dmitripivkine, okay I'm also updating `mmparse.cpp`, within a conditional 64-bit section so as to avoid enabling this for 32-bit versions of OpenJ9.

It's a bit weird that this seems to duplicate `vmargs.c`'s logic, yet in a different style...

File: runtime/gc_modron_startup/mmparse.cpp [hunk 1]
```diff
@@ -1376,6 +1376,15 @@ scan_udata_memory_size_helper(J9JavaVM *javaVM, char **cursor, UDATA *value, con
 			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTIONS_VALUE_OVERFLOWED, argName);
 			return false;
 		}
+#if defined(J9VM_ENV_DATA64)
+	} else if(try_scan(cursor, "T") || try_scan(cursor, "t")) {
```
Comment (dmitripivkine) [inline runtime/gc_modron_startup/mmparse.cpp]: please move this up to have scan in order T, G, M, K:
```
#if defined(J9VM_ENV_DATA64)
     if(try_scan(cursor, "T") || try_scan(cursor, "t")) {
.............
     } else
#endif /* defined(J9VM_ENV_DATA64) */
    if(try_scan(cursor, "G") || try_scan(cursor, "g")) {
```

Comment (avermeer) [reply runtime/gc_modron_startup/mmparse.cpp]: Okay will do

**Comment by avermeer** [comment]
I think I am going nuts with squashing: I follow instructions at https://blog.carbonfive.com/2017/08/28/always-squash-and-rebase-your-git-commits/ and then when I re-checkout my branch and run:
`git log --graph --decorate --pretty=oneline --abbrev-commit`
then I still see as many commits as before, just like if nothing was merged.

**Comment by avermeer** [comment]
Hi @dmitripivkine, I applied your suggested change. Any other feedback?

**Comment by pshipton** [comment]
@keithc-ca @DanHeidinga any advice on how to squash commits using the git command line?

**Comment by pshipton** [comment]
FWIW the earlier PR testing passed.
https://ci.eclipse.org/openj9/view/Pull%20Requests/job/PullRequest-OpenJ9/1842/

**Comment by avermeer** [comment]
Hello @pshipton, I found a friend who helped me with git rebasing (long story...), so normally I have now only one commit, and validation sounds good.

What is the next step to get this PR merged, and finally when could I see my fix in AdoptOpenJDK OpenJ9 nightly builds?

Kind regards,
Alexandre

**Comment by pshipton** [comment]
@avermeer there is still an extra commit attached to this PR "Fix freemarker.jar path in Build instructions for JDK 11 and 13", can you please remove it.

File: runtime/gc_modron_startup/mmparse.cpp (hunk 1 above)
Comment (pshipton) [reply runtime/gc_modron_startup/mmparse.cpp]: The code was copied up, but needs to be removed here.

Comment (pshipton) [reply runtime/gc_modron_startup/mmparse.cpp]: @dmitripivkine are there any tests for this GC code that need to be updated?

**Comment by pshipton** [comment]
> What is the next step to get this PR merged, and finally when could I see my fix in AdoptOpenJDK OpenJ9 nightly builds?

Once the requested changes are made, and the reviewers have approved, we run PR testing and it passes, then this can be merged. It won't happen before Thursday even if this is ready before then, as we have some platforms offline for maintenance (xlinux and mac) until Thursday. Once it is merged, it will become part of the next AdoptOpenJDK nightly build.

File: runtime/gc_modron_startup/mmparse.cpp (hunk 1 above)
Comment (dmitripivkine) [reply runtime/gc_modron_startup/mmparse.cpp]: > @dmitripivkine are there any tests for this GC code that need to be updated?

We do have implicit check in tests like https://github.com/eclipse/openj9/blob/master/test/functional/cmdLineTests/cmdLineTest_J9tests/j9tests_Java9.xml
or GC regression tests. We did have explicit parsing-only for `G`,`M`,`K` in the past but I can not find it now.
So I believe this change is good enough for now (duplicated code must be removed obviously)

**Comment by keithc-ca** [comment]
I wonder whether `T` should be recognized even on 32-bit platforms; although it likely would just yield some kind of 'overflow' response. What do others think?

File: runtime/gc_modron_startup/mmparse.cpp (hunk 1 above)
Comment (avermeer) [reply runtime/gc_modron_startup/mmparse.cpp]: Hello, I removed the extraneous lines from `mmparse.cpp` (yet another impact of my lack of habits with squashing, but I'm learning). Now I face another strange issue: the travis CI build is finished according to the Travis [details](https://travis-ci.org/eclipse/openj9/builds/610952004?utm_source=github_status&utm_medium=notification), but current PR page fails to acknowledge that  CI build is finished for a while - even if I refresh current page in my browse.

Did I hit yet another newbie issue with PR process?

Comment (avermeer) [reply runtime/gc_modron_startup/mmparse.cpp]: @pshipton any idea why the PR page still displays that Travis CI is in progress while on Travis details page it's done since hours?

Comment (pshipton) [reply runtime/gc_modron_startup/mmparse.cpp]: Sometimes there are hiccups. It won't block merging, you don't need to be concerned. The travis build has limited value, it only tests Java 11 compilation with cmake, it doesn't run any testing.

**Comment by pshipton** [comment]
> I wonder whether T should be recognized even on 32-bit platforms; although it likely would just yield some kind of 'overflow' response. 

Yes, doing this would match the Hotspot behavior.

**Comment by avermeer** [comment]
Okay I can add "t"/"T" support on 32-bit, but then would you please provide me with 32-bit build & test instructions ?

**Comment by pshipton** [comment]
> would you please provide me with 32-bit build & test instructions ?

The only 32-bit platform supported by OpenJ9 is Java 8 Windows. 
Example build https://ci.eclipse.org/openj9/job/Build_JDK8_x86-32_windows_Nightly/203/
Instructions https://github.com/eclipse/openj9/blob/master/doc/build-instructions/Build_Instructions_V8.md#windows

I can understand if you don't want to build/test this platform. If you make the code changes we can use PR testing to compile and test.

**Comment by avermeer** [comment]
@pshipton you read my mind: I do not do Windows :)

Here we go, I made the changes for 't' and 'T' suffixes to be supported on 32b-bits (with successful CI, now running agin since I forced push to merge my commits into one...)

I'll have to re-roll the change on documentation to remove limitation on 32-bit...

**Comment by pshipton** [comment]
We'll see if this compiles.
jenkins compile win32 jdk8

File: runtime/gc_modron_startup/mmparse.cpp [hunk 2]
```diff
@@ -1369,7 +1369,14 @@ scan_udata_memory_size_helper(J9JavaVM *javaVM, char **cursor, UDATA *value, con
 		return false;
 	}
 	
-	if(try_scan(cursor, "G") || try_scan(cursor, "g")) {
+	if(try_scan(cursor, "T") || try_scan(cursor, "t")) {
+		if (*value <= (((UDATA)-1) >> 40)) {
+			*value <<= 40;
+		} else {
```
Comment (keithc-ca) [inline runtime/gc_modron_startup/mmparse.cpp]: I expect the 32-bit windows compile will fail here (shifting by amounts greater than the width of `UDATA`).

Comment (pshipton) [reply runtime/gc_modron_startup/mmparse.cpp]: @avermeer given Keith's comment, let's use an #ifdef to exclude the bit shift code on 32-bit. Even if it compiles on Windows 32-bit, the code does successfully compile for 32-bit on Java 8 for most platforms and we'd like to keep it that way, although they aren't officially supported or testable at the OpenJ9 project.

Comment (dmitripivkine) [reply runtime/gc_modron_startup/mmparse.cpp]: > I expect the 32-bit windows compile will fail here (shifting by amounts greater than the width of `UDATA`).

As I remember it does mod32 (or circular) shift, so it is shift by 8 most likely

Comment (dmitripivkine) [reply runtime/gc_modron_startup/mmparse.cpp]: I believe you can do something like this:
``` 
       if(try_scan(cursor, "T") || try_scan(cursor, "t")) {
#if defined(J9VM_ENV_DATA64)
		if (*value <= (((UDATA)-1) >> 40)) {
 			*value <<= 40;
 		} else
#endif /* defined(J9VM_ENV_DATA64) */ 
               {
			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTIONS_VALUE_OVERFLOWED, argName);
 			return false;
 		}
        } else if(try_scan(cursor, "G") || try_scan(cursor, "g")) {
```

**Comment by pshipton** [comment]
The compile did fail. It's only a warning, but we treat warnings as errors in this code.
```
17:45:23  vmargs.c(613) : error C2220: warning treated as error - no 'object' file generated
17:45:23  vmargs.c(613) : warning C4293: '>>' : shift count negative or too big, undefined behavior
```

**Comment by keithc-ca** [comment]
Could you please fix the 'Author' of your commit(s); it should mention your name instead of 'root'.
Within the working tree of your clone, you can use these commands to do that:
```
git config --global user.name 'Alexandre Vermeerbergen'
git config --global user.email 'avermeerbergen@gmail.com'
git commit --amend --reset-author
```
Omit `--global` if you want the configuration changes to be applied only to the specific clone.

File: runtime/gc_modron_startup/mmparse.cpp (hunk 2 above)
Comment (pshipton) [reply runtime/gc_modron_startup/mmparse.cpp]: Yes, it is technically correct to support zero (`0T`) in the parsing functions.

**Comment by avermeer** [comment]
Okay let me get back on this PR will all above remarks... I'll update soon

File: runtime/util/vmargs.c [hunk 3]
```diff
@@ -608,6 +608,21 @@ optionValueOperations(J9PortLibrary *portLibrary, J9VMInitArgs* j9vm_args, IDATA
...
 							break;
+						case 't':
+						case 'T':
+#if defined(J9VM_ENV_DATA64)
+							if (value <= (((UDATA)-1) >> 40)) {
+								value <<= 40;
+							} else {
+								return OPTION_OVERFLOW;
+							}
+#else /* !defined(J9VM_ENV_DATA64) */
```
Comment (keithc-ca) [inline runtime/util/vmargs.c]: By convention, we repeat the `#if` condition (not it's negation) here.

File: runtime/gc_modron_startup/mmparse.cpp [hunk 4]
```diff
@@ -1369,7 +1369,19 @@ scan_udata_memory_size_helper(J9JavaVM *javaVM, char **cursor, UDATA *value, con
 		return false;
 	}
 	
-	if(try_scan(cursor, "G") || try_scan(cursor, "g")) {
+	if(try_scan(cursor, "T") || try_scan(cursor, "t")) {
+		if (0 != value) {
```
Comment (keithc-ca) [inline runtime/gc_modron_startup/mmparse.cpp]: I expect you meant `(0 != *value)` (to dereference `value`).

Comment (avermeer) [reply runtime/gc_modron_startup/mmparse.cpp]: oops, my bad, let me fix that

File: runtime/vm/jvminit.c [hunk 5]
```diff
@@ -5175,6 +5175,50 @@ static void testOptionValueOps(J9JavaVM* vm) {
...
+	SET_TO(1, "-Xfoi8388608T");			/* 2^63+1 */
+	optName = "-Xfoi";
+	intResult = GET_MEMORY_VALUE(1, optName, uResult);
+	TEST_INT(intResult, OPTION_OVERFLOW);
+
+	SET_TO(1, "-Xfoj8388608t");			/* 2^63+1 */
+	optName = "-Xfoj";
+	intResult = GET_MEMORY_VALUE(1, optName, uResult);
+	TEST_INT(intResult, OPTION_OVERFLOW);
+#else	/* !defined(J9VM_ENV_DATA64) */
```
Comment (pshipton) [inline runtime/vm/jvminit.c]: This comment should also be fixed.

**Comment by pshipton** [comment]
jenkins test sanity win32 jdk8

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/PullRequest-OpenJ9/1862/

File: runtime/vm/jvminit.c (hunk 5 above)
Comment (avermeer) [reply runtime/vm/jvminit.c]: @pshipton , alright, fixed the comment in `jvminit.c`.

I have some questions:
1.  The size suffix parsing logic seems to be duplicated in `vmargs.c` and in `mmparse.cpp`, as pointed out by @dmitripivkine.
Is there a reason for such duplication?
Which one is used for which case?
2. When I un-commented the `#define JVMINIT_UNIT_TEST` in `jvminit`.c, I got some failed tests, even before changing it.
When I added the unit tests for overflow cases, I see I added more failed tests, in spite of my effort to detect that the overflow error was expected.
I would love to get more insights on this area, because I am feeling uncomfortable to get a test summary result being "Failed", even before my changes.
Maybe should I open another incident to track this unit testing issue, as it's kind of independent of current issue?

File: runtime/vm/jvminit.c [hunk 6]
```diff
@@ -5175,6 +5175,50 @@ static void testOptionValueOps(J9JavaVM* vm) {
 	TEST_INT(uResult, (1024*1024*1024));
 	TEST_INT(intResult, OPTION_OK);
 
+#if defined(J9VM_ENV_DATA64)
+	SET_TO(1, "-Xfog8388607T");			/* 2^63 */
```
Comment (keithc-ca) [inline runtime/vm/jvminit.c]: Shouldn't the boundary be at 16777216T (2^24 T), meaning 16777215T should be accepted?
The code uses UDATA which (on 64-bit platforms) holds values up to 2^64-1. After shifting right 40 bits, the maximum value should be 2^24-1.

File: runtime/vm/jvminit.c (hunk 5 above)
Comment (pshipton) [reply runtime/vm/jvminit.c]: Yes, you can open an Issue for any failures that occur previous to your change.

The `mmparse.cpp` code is used to parse options that affect the object heap sizes, like `-Xmx` or `-XX:MaxHeapSize`. The reason for having duplicate code is historic, as the code bases used to be separated.

File: runtime/vm/jvminit.c (hunk 6 above)
Comment (avermeer) [reply runtime/vm/jvminit.c]: @keithc-ca you were right, `16777215T `(=`2^24-1` indeed) is the maximum size that `vmarg.c` can accept with 'T'/'t' suffix.  My unit tests  on 't'/'T' are now all in "Succeeded" status, thank you very much for your review!

**Comment by keithc-ca** [comment]
Jenkins test sanity win,win32 jdk8

File: runtime/vm/jvminit.c (hunk 5 above)
Comment (avermeer) [reply runtime/vm/jvminit.c]: @keithc-ca I have included all your remarks in my latest commit. Please let me know if there is anything else that's missing to get this PR integrated.

**Comment by pshipton** [comment]
@avermeer congratulations on your first commit!

**Comment by avermeer** [comment]
@pshipton thank you very much for your guidance in this process. I will soon push the latest documentation update related to this change, and of course, I stay tuned to get the 1st nightly build that integrate this PR.
